pkg/ruler/ruler.go

```diff
@@ -430,18 +449,147 @@ func (r *Ruler) newManager(ctx context.Context, userID string) (*promRules.Manag
 		ExternalURL: r.alertURL,
 		NotifyFunc:  sendAlerts(notifier, r.alertURL.String()),
-		Logger:      util.Logger,
+		Logger:      logger,
 		Registerer:  reg,
 	}
 	return promRules.NewManager(opts), nil
 }
+
+// GetRules retrieves the running rules from this ruler and all running rulers in the ring if
+// sharding is enabled
+func (r *Ruler) GetRules(ctx context.Context, userID string) ([]*rules.RuleGroupDesc, error) {
+	if r.cfg.EnableSharding {
+		return r.getShardedRules(ctx, userID)
+	}
+
+	return r.getLocalRules(userID)
+}
+
+func (r *Ruler) getLocalRules(userID string) ([]*rules.RuleGroupDesc, error) {
+	groupDescs := []*rules.RuleGroupDesc{}
+
+	var groups []*promRules.Group
+	r.userManagerMtx.Lock()
+	mngr, exists := r.userManagers[userID]
+	if exists {
+		groups = mngr.RuleGroups()
+	}
+	r.userManagerMtx.Unlock()
+
+	prefix := filepath.Join(r.cfg.RulePath, userID) + "/"
+
+	for _, group := range groups {
+		interval := group.Interval()
+		groupDesc := &rules.RuleGroupDesc{
+			Name:      group.Name(),
+			Namespace: strings.TrimPrefix(group.File(), prefix),
+			Interval:  &interval,
+			User:      userID,
+		}
+		for _, r := range group.Rules() {
+			lastError := ""
+			if r.LastError() != nil {
+				lastError = r.LastError().Error()
+			}
+
+			var ruleDesc *rules.RuleDesc
+			switch rule := r.(type) {
+			case *promRules.AlertingRule:
+				dur := rule.Duration()
+				rule.ActiveAlerts()
+				alerts := []*rules.Alert{}
+				for _, a := range rule.ActiveAlerts() {
+					alerts = append(alerts, &rules.Alert{
+						State:       a.State.String(),
+						Labels:      client.FromLabelsToLabelAdapters(a.Labels),
+						Annotations: client.FromLabelsToLabelAdapters(a.Annotations),
+						Value:       a.Value,
+						ActiveAt:    &a.ActiveAt,
+						FiredAt:     &a.FiredAt,
+						ResolvedAt:  &a.ResolvedAt,
+						LastSentAt:  &a.LastSentAt,
+						ValidUntil:  &a.ValidUntil,
+					})
+				}
+				ruleDesc = &rules.RuleDesc{
+					State:       rule.State().String(),
+					Alert:       rule.Name(),
+					Alerts:      alerts,
+					Expr:        rule.Query().String(),
+					For:         &dur,
+					Labels:      client.FromLabelsToLabelAdapters(rule.Labels()),
+					Annotations: client.FromLabelsToLabelAdapters(rule.Annotations()),
+					Health:      string(rule.Health()),
+					LastError:   lastError,
+				}
+			case *promRules.RecordingRule:
+				ruleDesc = &rules.RuleDesc{
+					Record:    rule.Name(),
+					Expr:      rule.Query().String(),
+					Labels:    client.FromLabelsToLabelAdapters(rule.Labels()),
+					Health:    string(rule.Health()),
+					LastError: lastError,
+				}
+			default:
+				return nil, errors.Errorf("failed to assert type of rule '%v'", rule.Name())
+			}
+			groupDesc.Rules = append(groupDesc.Rules, ruleDesc)
+		}
+		groupDescs = append(groupDescs, groupDesc)
+	}
+	return groupDescs, nil
+}
+
+func (r *Ruler) getShardedRules(ctx context.Context, userID string) ([]*rules.RuleGroupDesc, error) {
```

Comment:
I haven't manually tested it. Have you tested it with the latest version of the code? If not, may you do it, please?

Reply:
Not yet, will do.